stdlib/public/core/Array.swift

```diff
@@ -1865,3 +1865,168 @@ internal struct _ArrayAnyHashableBox<Element: Hashable>
+    }
+
+    @inlinable
+    public func moved(along direction: TangentVector) -> [Element.TangentVector] {
+      return zip(elements, direction.elements).map { $0.moved(along: $1) }
+    }
+
+    @inlinable
+    public func tangentVector(from cotangent: CotangentVector) -> TangentVector {
+      let elements = zip(self.elements, cotangent.elements).map {
+        $0.tangentVector(from: $1)
+      }
+      return TangentVector(elements: elements)
+    }
+  }
+
+  public struct CotangentVector : AdditiveArithmetic & Differentiable {
+    public typealias TangentVector = Array<Element>.CotangentVector
+    public typealias CotangentVector = Array<Element>.TangentVector
+
+    public var elements: [Element.CotangentVector]
+
+    public static var zero: TangentVector {
+      return TangentVector(elements: [])
+    }
+    
+    public init(elements: [Element.CotangentVector]) {
+      self.elements = elements
+    }
+
+    @inlinable
+    public static func + (lhs: TangentVector, rhs: TangentVector) -> TangentVector {
+      return TangentVector(elements: zipLongest(lhs.elements, rhs.elements).map {
+        ($0 ?? .zero) + ($1 ?? .zero)
+      })
+    }
+
+    @inlinable
+    public static func - (lhs: TangentVector, rhs: TangentVector) -> TangentVector {
+      return TangentVector(elements: zipLongest(lhs.elements, rhs.elements).map {
+        ($0 ?? .zero) - ($1 ?? .zero)
+      })
+    }
+
+    @inlinable
+    public func moved(along direction: TangentVector) -> [Element.CotangentVector] {
+      return zip(elements, direction.elements).map { $0.moved(along: $1) }
+    }
+
+    @inlinable
+    public func tangentVector(from cotangent: CotangentVector) -> TangentVector {
+      let elements = zip(self.elements, cotangent.elements).map {
+        $0.tangentVector(from: $1)
+      }
+      return TangentVector(elements: elements)
+    }
+  }
+
+  public typealias AllDifferentiableVariables = [Element]
+
+  @inlinable
+  public func moved(along direction: TangentVector) -> [Element] {
+    return zip(self, direction.elements).map { $0.moved(along: $1) }
+  }
+
+  @inlinable
+  public func tangentVector(from cotangent: CotangentVector) -> TangentVector {
+    return TangentVector(elements: zip(self, cotangent.elements).map { $0.tangentVector(from: $1) })
+  }
+}
+
+// TODO: Remove once `zipLongest` is part of stdlib.
+public func zipLongest<Sequence1: Sequence, Sequence2: Sequence>(
+  _ sequence1: Sequence1, 
+  _ sequence2: Sequence2
+) -> [(Sequence1.Element?, Sequence2.Element?)] {
+  var (iterator1, iterator2) = (sequence1.makeIterator(), sequence2.makeIterator())
+  var result = [(Sequence1.Element?, Sequence2.Element?)]()
+  
+  result.reserveCapacity(Swift.max(
+    sequence1.underestimatedCount, 
+    sequence2.underestimatedCount))
+
+  while true {
+    let (element1, element2) = (iterator1.next(), iterator2.next())
+    if element1 == nil && element2 == nil {
+      break
+    } else {
+      result.append((element1, element2))
```

Comment:
Since the `if` body would break, how about moving the `append` statement out of the `else` block?